sensor/tests/resync_test.go

```diff
@@ -0,0 +1,187 @@
+	utils.CrashOnError(os.Setenv("ROX_MTLS_CA_KEY_FILE", "../../tools/local-sensor/certs/caKey.pem"))
+
+	fakeCentral := centralDebug.MakeFakeCentralWithInitialMessages(
+		message.SensorHello("1234"),
+		message.ClusterConfig(),
+		message.PolicySync([]*storage.Policy{}),
+		message.BaselineSync([]*storage.ProcessBaseline{}))
+
+	conn, spyCentral, shutdownFakeServer := createConnectionAndStartServer(fakeCentral)
+	defer shutdownFakeServer()
+	fakeConnectionFactory := centralDebug.MakeFakeConnectionFactory(conn)
+
+	s, err := sensor.CreateSensor(sensor.ConfigWithDefaults().
+		WithK8sClient(fakeClient).
+		WithLocalSensor(true).
+		WithResyncPeriod(1 * time.Second).
+		WithCentralConnectionFactory(fakeConnectionFactory))
+
+	if err != nil {
+		panic(err)
+	}
+
+	go s.Start()
+	defer s.Stop()
+
+	spyCentral.ConnectionStarted.Wait()
+
+	testCases := map[string]struct {
+		orderedEvents  []func(t *testing.T, k *k8s.ClientSet)
+		deploymentName string
+	}{
+		// This should not work if re-sync is disabled, because deployment will be
+		// sent to central before RBAC is computed fully.
+		"Deployment first": {
+			orderedEvents: []func(t *testing.T, k *k8s.ClientSet){
+				CreateDeployment("dep1", "sa1"),
+				CreateRole("r1"),
+				CreateRoleBinding("b1", "r1", "sa1"),
+			},
+			deploymentName: "dep1",
+		},
+		// This should also not work, if just the role or just the bindings is
+		// available, the correct permission level won't be determined correctly.
+		"Deployment second": {
+			orderedEvents: []func(t *testing.T, k *k8s.ClientSet){
+				CreateRole("r2"),
+				CreateDeployment("dep2", "sa2"),
+				CreateRoleBinding("b2", "r2", "sa2"),
+			},
+			deploymentName: "dep2",
+		},
+		// This is the only case that should work if re-sync isn't enabled.
+		"Deployment last": {
+			orderedEvents: []func(t *testing.T, k *k8s.ClientSet){
+				CreateRole("r3"),
+				CreateRoleBinding("b3", "r3", "sa3"),
+				CreateDeployment("dep3", "sa3"),
+			},
+			deploymentName: "dep3",
+		},
+	}
+
+	fakeClient.SetupExampleCluster(t)
+	for name, testCase := range testCases {
+		t.Run(name, func(t *testing.T) {
+			for _, fn := range testCase.orderedEvents {
+				// This function will create the fake k8s event and wait for one second.
+				// This way we guarantee some level of order for the events. Since sensor
+				// processes events in different goroutines, some events take more time to
+				// be processed than others. Therefore, even if we send two different events
+				// at a particular order, e.g. Deployment and role respectively, it might be
+				// that Role finishes first and is sent to Central before deployment gets to
+				// the point of checking for permission levels.
+				// To force what would happen in conditions where deployment is fully processed
+				// before a Role is received, a timer was introduced here as well.
+				// TODO: This should be done by checking the received event rather than arbitrarily waiting.
```

Comment:
Should we create an issue for this?

Reply:
Sure. If we're okay not having this for now and working on it in parallel then I'll create an issue for it.

Reply:
Yea, I tried a couple of times yesterday and wasn't able to get this logic working. I tried it now and works like a charm. Maybe I messed something up in my original implementations. 🤷 
Pushing the new version now. Thanks!